File: src/daemon/routes/policy.ts
/**
 * Author: Deepankar Das
 */

/**
 * Enforcer — Policy Management API
 *
 * REST endpoints for policy CRUD and hot reload:
 *   GET    /v1/policy/rules           — list all active rules
 *   POST   /v1/policy/rules           — add a new rule
 *   PUT    /v1/policy/rules/:id       — update a rule
 *   DELETE /v1/policy/rules/:id       — delete a rule
 *   POST   /v1/policy/rules/:id/toggle — enable/disable a rule
 *   GET    /v1/policy/bundle          — get current bundle metadata
 *   POST   /v1/policy/reload          — reload from disk
 *   GET    /v1/policy/packs           — list available canned packs
 *   POST   /v1/policy/packs/:id/apply — apply a canned pack
 */

import type { PolicyBundle, PolicyRule } from "../../../types/policy.js";
import { getAvailablePacks, getPack, applyPack, addCustomPack } from "../../policy/packs.js";

export interface PolicyRouteResult {
  status: number;
  body: unknown;
}

// ── In-memory state for active policy bundle ────────────────────────────────
// The daemon passes its policyBundle reference here. Mutations are live.

let activeBundleRef: { bundle: PolicyBundle } | null = null;

export function setPolicyBundleRef(ref: { bundle: PolicyBundle }): void {
  activeBundleRef = ref;
}

// ── Track disabled rules ────────────────────────────────────────────────────

const disabledRules = new Set<string>();

export function isRuleEnabled(ruleId: string): boolean {
  return !disabledRules.has(ruleId);
}

// ── GET /v1/policy/rules ────────────────────────────────────────────────────

export function handleListRules(): PolicyRouteResult {
  if (!activeBundleRef) {
    return { status: 500, body: { error: "No policy bundle loaded" } };
  }

  const rules = activeBundleRef.bundle.rules.map(rule => ({
    ...rule,
    enabled: isRuleEnabled(rule.policy_id),
  }));

  return {
    status: 200,
    body: {
      bundle_version: activeBundleRef.bundle.bundle_version,
      rules,
      count: rules.length,
      enabled_count: rules.filter(r => r.enabled).length,
    },
  };
}

// ── POST /v1/policy/rules ───────────────────────────────────────────────────

export function handleAddRule(rawBody: string): PolicyRouteResult {
  if (!activeBundleRef) {
    return { status: 500, body: { error: "No policy bundle loaded" } };
  }

  let body: Record<string, unknown>;
  try {
    body = JSON.parse(rawBody);
  } catch {
    return { status: 400, body: { error: "Invalid JSON" } };
  }

  if (!body.policy_id || !body.action || !body.effect) {
    return { status: 400, body: { error: "policy_id, action, and effect are required" } };
  }

  // Build rule from input
  const rule: PolicyRule = {
    policy_id: body.policy_id as string,
    version: activeBundleRef.bundle.bundle_version,
    scope: { level: (body.scope_level as "organization" | "team" | "repository" | "local") || "organization" },
    subject: {
      agent_types: (body.agent_types as string[]) || ["*"],
      users: (body.users as string[]) || ["*"],
    },
    action: { types: (body.action as { types: string[] }).types || [] },
    resource: (body.resource as Record<string, unknown>) || {},
    conditions: (body.conditions as Record<string, unknown>) || {},
    effect: body.effect as { decision: "allow" | "deny" | "require_approval"; reason_code: string; reason_human: string },
    logging: { mode: "full" },
    approval: { required: (body.effect as { decision: string }).decision === "require_approval" },
  };

  // Check for duplicate policy_id
  if (activeBundleRef.bundle.rules.some(r => r.policy_id === rule.policy_id)) {
    return { status: 409, body: { error: `Rule ${rule.policy_id} already exists. Use PUT to update.` } };
  }

  activeBundleRef.bundle.rules.push(rule);

  return {
    status: 201,
    body: { message: `Rule ${rule.policy_id} added`, rule },
  };
}

// ── PUT /v1/policy/rules/:id ────────────────────────────────────────────────

export function handleUpdateRule(ruleId: string, rawBody: string): PolicyRouteResult {
  if (!activeBundleRef) {
    return { status: 500, body: { error: "No policy bundle loaded" } };
  }

  const index = activeBundleRef.bundle.rules.findIndex(r => r.policy_id === ruleId);
  if (index === -1) {
    return { status: 404, body: { error: `Rule ${ruleId} not found` } };
  }

  let body: Record<string, unknown>;
  try {
    body = JSON.parse(rawBody);
  } catch {
    return { status: 400, body: { error: "Invalid JSON" } };
  }

  const existing = activeBundleRef.bundle.rules[index];

  // Merge updates
  if (body.action) existing.action = body.action as { types: string[] };
  if (body.resource) existing.resource = body.resource as Record<string, unknown>;
  if (body.conditions) existing.conditions = body.conditions as Record<string, unknown>;
  if (body.effect) {
    existing.effect = body.effect as { decision: "allow" | "deny" | "require_approval"; reason_code: string; reason_human: string };
    existing.approval = { required: existing.effect.decision === "require_approval" };
  }
  if (body.subject) existing.subject = body.subject as { agent_types: string[]; users: string[] };

  return {
    status: 200,
    body: { message: `Rule ${ruleId} updated`, rule: existing },
  };
}

// ── DELETE /v1/policy/rules/:id ─────────────────────────────────────────────

export function handleDeleteRule(ruleId: string): PolicyRouteResult {
  if (!activeBundleRef) {
    return { status: 500, body: { error: "No policy bundle loaded" } };
  }

  const index = activeBundleRef.bundle.rules.findIndex(r => r.policy_id === ruleId);
  if (index === -1) {
    return { status: 404, body: { error: `Rule ${ruleId} not found` } };
  }

  const removed = activeBundleRef.bundle.rules.splice(index, 1)[0];
  disabledRules.delete(ruleId);

  return {
    status: 200,
    body: { message: `Rule ${ruleId} deleted`, rule: removed },
  };
}

// ── POST /v1/policy/rules/:id/toggle ────────────────────────────────────────

export function handleToggleRule(ruleId: string): PolicyRouteResult {
  if (!activeBundleRef) {
    return { status: 500, body: { error: "No policy bundle loaded" } };
  }

  const rule = activeBundleRef.bundle.rules.find(r => r.policy_id === ruleId);
  if (!rule) {
    return { status: 404, body: { error: `Rule ${ruleId} not found` } };
  }

  if (disabledRules.has(ruleId)) {
    disabledRules.delete(ruleId);
    return { status: 200, body: { policy_id: ruleId, enabled: true, message: `Rule ${ruleId} enabled` } };
  } else {
    disabledRules.add(ruleId);
    return { status: 200, body: { policy_id: ruleId, enabled: false, message: `Rule ${ruleId} disabled` } };
  }
}

// ── GET /v1/policy/bundle ───────────────────────────────────────────────────

export function handleGetBundle(): PolicyRouteResult {
  if (!activeBundleRef) {
    return { status: 500, body: { error: "No policy bundle loaded" } };
  }

  return {
    status: 200,
    body: {
      bundle_version: activeBundleRef.bundle.bundle_version,
      scope_level: activeBundleRef.bundle.scope_level,
      rule_count: activeBundleRef.bundle.rules.length,
      enabled_count: activeBundleRef.bundle.rules.filter(r => isRuleEnabled(r.policy_id)).length,
      disabled_count: disabledRules.size,
    },
  };
}

// ── GET /v1/policy/packs ────────────────────────────────────────────────────

export function handleListPacks(): PolicyRouteResult {
  const packs = getAvailablePacks().map(pack => ({
    id: pack.id,
    name: pack.name,
    description: pack.description,
    category: pack.category,
    tags: pack.tags,
    rule_count: pack.rules.length,
  }));

  return {
    status: 200,
    body: { packs, count: packs.length },
  };
}

// ── POST /v1/policy/packs/:id/apply ─────────────────────────────────────────

export function handleApplyPack(packId: string): PolicyRouteResult {
  if (!activeBundleRef) {
    return { status: 500, body: { error: "No policy bundle loaded" } };
  }

  try {
    const result = applyPack(activeBundleRef.bundle, packId);
    return {
      status: 200,
      body: {
        message: `Pack ${packId} applied`,
        added: result.added,
        skipped: result.skipped,
        total_rules: activeBundleRef.bundle.rules.length,
      },
    };
  } catch (err) {
    const message = err instanceof Error ? err.message : "Unknown error";
    return { status: 404, body: { error: message } };
  }
}

// ── GET /v1/policy/packs/:id ────────────────────────────────────────────────

export function handleGetPack(packId: string): PolicyRouteResult {
  const pack = getPack(packId);
  if (!pack) {
    return { status: 404, body: { error: `Pack ${packId} not found` } };
  }

  return {
    status: 200,
    body: {
      id: pack.id,
      name: pack.name,
      description: pack.description,
      category: pack.category,
      tags: pack.tags,
      rules: pack.rules.map(r => ({
        policy_id: r.policy_id,
        action_types: r.action.types,
        decision: r.effect.decision,
        reason_code: r.effect.reason_code,
        reason_human: r.effect.reason_human,
      })),
      rule_count: pack.rules.length,
    },
  };
}

// ── POST /v1/policy/packs ───────────────────────────────────────────────────
// Create a custom pack from a list of rule IDs (copies from active bundle)
// or from provided rules directly.

export function handleCreatePack(rawBody: string): PolicyRouteResult {
  let body: Record<string, unknown>;
  try {
    body = JSON.parse(rawBody);
  } catch {
    return { status: 400, body: { error: "Invalid JSON" } };
  }

  if (!body.id || !body.name) {
    return { status: 400, body: { error: "id and name are required" } };
  }

  const packId = body.id as string;
  const packName = body.name as string;
  const description = (body.description as string) || "";
  const category = (body.category as string) || "Custom";
  const tags = (body.tags as string[]) || ["custom"];

  // Option 1: Copy rules from active bundle by rule IDs
  const ruleIds = body.rule_ids as string[] | undefined;
  // Option 2: Use a template pack and customize
  const templatePackId = body.template_pack_id as string | undefined;
  // Option 3: Provide rules directly
  const directRules = body.rules as Array<{
    policy_id: string;
    action_types: string[];
    decision: string;
    reason_code: string;
    reason_human: string;
  }> | undefined;

  let rules: PolicyRule[] = [];

  if (templatePackId) {
    const template = getPack(templatePackId);
    if (!template) {
      return { status: 404, body: { error: `Template pack ${templatePackId} not found` } };
    }
    // Clone template rules with new pack prefix
    rules = template.rules.map(r => ({
      ...r,
      policy_id: r.policy_id.replace(/^pack\./, `custom.${packId}.`),
    }));
  } else if (ruleIds && activeBundleRef) {
    for (const id of ruleIds) {
      const rule = activeBundleRef.bundle.rules.find(r => r.policy_id === id);
      if (rule) {
        rules.push({ ...rule, policy_id: `custom.${packId}.${rule.policy_id}` });
      }
    }
  } else if (directRules) {
    rules = directRules.map(r => ({
      policy_id: r.policy_id,
      version: activeBundleRef?.bundle.bundle_version || "v1",
      scope: { level: "organization" as const },
      subject: { agent_types: ["*"], users: ["*"] },
      action: { types: r.action_types },
      resource: {},
      conditions: {},
      effect: {
        decision: r.decision as "allow" | "deny" | "require_approval",
        reason_code: r.reason_code,
        reason_human: r.reason_human,
      },
      logging: { mode: "full" as const },
      approval: { required: r.decision === "require_approval" },
    }));
  }

  if (rules.length === 0) {
    return { status: 400, body: { error: "No rules provided. Use rule_ids, template_pack_id, or rules." } };
  }

  try {
    addCustomPack({
      id: packId,
      name: packName,
      description,
      category,
      tags,
      rules,
    });

    return {
      status: 201,
      body: {
        message: `Pack ${packId} created with ${rules.length} rules`,
        id: packId,
        rule_count: rules.length,
      },
    };
  } catch (err) {
    const msg = err instanceof Error ? err.message : "Unknown error";
    return { status: 409, body: { error: msg } };
  }
}

File: go/internal/daemon/routes/policy.go
/**
 * Author: Deepankar Das
 */

package routes

import (
	"encoding/json"
	"net/http"
	"sync"

	"github.com/anthropics/enforcer/internal/policy"
	"github.com/anthropics/enforcer/internal/types"
)

// PolicyRoutes holds a reference to the live policy bundle (protected by a
// read-write mutex) and exposes handler methods for policy CRUD operations.
type PolicyRoutes struct {
	sync.RWMutex
	bundle       *types.PolicyBundle
	saveBundleFn func(types.PolicyBundle) error
	savePacksFn  func([]policy.PolicyPack) error
}

// NewPolicyRoutes creates a PolicyRoutes wrapping the given bundle.
func NewPolicyRoutes(bundle *types.PolicyBundle, saveBundleFn func(types.PolicyBundle) error, savePacksFn func([]policy.PolicyPack) error) *PolicyRoutes {
	return &PolicyRoutes{
		bundle:       bundle,
		saveBundleFn: saveBundleFn,
		savePacksFn:  savePacksFn,
	}
}

// GetBundle returns a copy of the current bundle.
func (pr *PolicyRoutes) GetBundle() types.PolicyBundle {
	return *pr.bundle
}

// --- Rules CRUD ---

// HandleListRules returns all rules in the bundle.
func (pr *PolicyRoutes) HandleListRules() (int, interface{}) {
	pr.RLock()
	defer pr.RUnlock()

	return http.StatusOK, map[string]interface{}{
		"rules": pr.bundle.Rules,
		"count": len(pr.bundle.Rules),
	}
}

// HandleAddRule adds a new rule to the bundle.
func (pr *PolicyRoutes) HandleAddRule(body []byte) (int, interface{}) {
	var rule types.PolicyRule
	if err := json.Unmarshal(body, &rule); err != nil {
		return http.StatusBadRequest, map[string]string{
			"error":   "invalid_json",
			"message": "Failed to parse policy rule: " + err.Error(),
		}
	}

	if rule.PolicyID == "" {
		return http.StatusBadRequest, map[string]string{
			"error":   "missing_policy_id",
			"message": "policy_id is required.",
		}
	}

	pr.Lock()
	defer pr.Unlock()

	// Check for duplicate policy_id.
	for _, existing := range pr.bundle.Rules {
		if existing.PolicyID == rule.PolicyID {
			return http.StatusConflict, map[string]string{
				"error":   "duplicate_policy_id",
				"message": "A rule with policy_id '" + rule.PolicyID + "' already exists.",
			}
		}
	}

	pr.bundle.Rules = append(pr.bundle.Rules, rule)
	if pr.saveBundleFn != nil {
		if err := pr.saveBundleFn(*pr.bundle); err != nil {
			return http.StatusInternalServerError, map[string]string{
				"error":   "persist_failed",
				"message": "Rule added in memory but failed to persist: " + err.Error(),
			}
		}
	}

	return http.StatusCreated, map[string]interface{}{
		"policy_id": rule.PolicyID,
		"added":     true,
		"total":     len(pr.bundle.Rules),
	}
}

// HandleUpdateRule replaces an existing rule by policy_id.
func (pr *PolicyRoutes) HandleUpdateRule(ruleID string, body []byte) (int, interface{}) {
	var rule types.PolicyRule
	if err := json.Unmarshal(body, &rule); err != nil {
		return http.StatusBadRequest, map[string]string{
			"error":   "invalid_json",
			"message": "Failed to parse policy rule: " + err.Error(),
		}
	}

	pr.Lock()
	defer pr.Unlock()

	for i, existing := range pr.bundle.Rules {
		if existing.PolicyID == ruleID {
			rule.PolicyID = ruleID // Preserve original ID.
			pr.bundle.Rules[i] = rule
			if pr.saveBundleFn != nil {
				if err := pr.saveBundleFn(*pr.bundle); err != nil {
					return http.StatusInternalServerError, map[string]string{
						"error":   "persist_failed",
						"message": "Rule updated in memory but failed to persist: " + err.Error(),
					}
				}
			}
			return http.StatusOK, map[string]interface{}{
				"policy_id": ruleID,
				"updated":   true,
			}
		}
	}

	return http.StatusNotFound, map[string]string{
		"error":   "rule_not_found",
		"message": "No rule with policy_id '" + ruleID + "' found.",
	}
}

// HandleDeleteRule removes a rule by policy_id.
func (pr *PolicyRoutes) HandleDeleteRule(ruleID string) (int, interface{}) {
	pr.Lock()
	defer pr.Unlock()

	for i, existing := range pr.bundle.Rules {
		if existing.PolicyID == ruleID {
			pr.bundle.Rules = append(pr.bundle.Rules[:i], pr.bundle.Rules[i+1:]...)
			if pr.saveBundleFn != nil {
				if err := pr.saveBundleFn(*pr.bundle); err != nil {
					return http.StatusInternalServerError, map[string]string{
						"error":   "persist_failed",
						"message": "Rule deleted in memory but failed to persist: " + err.Error(),
					}
				}
			}
			return http.StatusOK, map[string]interface{}{
				"policy_id": ruleID,
				"deleted":   true,
				"remaining": len(pr.bundle.Rules),
			}
		}
	}

	return http.StatusNotFound, map[string]string{
		"error":   "rule_not_found",
		"message": "No rule with policy_id '" + ruleID + "' found.",
	}
}

// HandleToggleRule toggles the enabled state of a rule by policy_id.
func (pr *PolicyRoutes) HandleToggleRule(ruleID string) (int, interface{}) {
	pr.Lock()
	defer pr.Unlock()

	for i, existing := range pr.bundle.Rules {
		if existing.PolicyID == ruleID {
			currentEnabled := existing.IsEnabled()
			newEnabled := !currentEnabled
			pr.bundle.Rules[i].Enabled = &newEnabled
			if pr.saveBundleFn != nil {
				if err := pr.saveBundleFn(*pr.bundle); err != nil {
					return http.StatusInternalServerError, map[string]string{
						"error":   "persist_failed",
						"message": "Rule toggled in memory but failed to persist: " + err.Error(),
					}
				}
			}
			return http.StatusOK, map[string]interface{}{
				"policy_id": ruleID,
				"enabled":   newEnabled,
			}
		}
	}

	return http.StatusNotFound, map[string]string{
		"error":   "rule_not_found",
		"message": "No rule with policy_id '" + ruleID + "' found.",
	}
}

// --- Bundle ---

// HandleGetBundle returns the full policy bundle.
func (pr *PolicyRoutes) HandleGetBundle() (int, interface{}) {
	pr.RLock()
	defer pr.RUnlock()

	return http.StatusOK, pr.bundle
}

// --- Packs ---

// HandleListPacks returns all available policy packs.
func (pr *PolicyRoutes) HandleListPacks() (int, interface{}) {
	packs := policy.GetAvailablePacks()
	return http.StatusOK, map[string]interface{}{
		"packs": packs,
		"count": len(packs),
	}
}

// HandleGetPack returns a single policy pack by ID.
func (pr *PolicyRoutes) HandleGetPack(packID string) (int, interface{}) {
	if packID == "" {
		return http.StatusBadRequest, map[string]string{
			"error":   "missing_pack_id",
			"message": "Pack ID is required.",
		}
	}

	pack := policy.GetPack(packID)
	if pack == nil {
		return http.StatusNotFound, map[string]string{
			"error":   "pack_not_found",
			"message": "No policy pack with ID '" + packID + "' found.",
		}
	}

	return http.StatusOK, pack
}

// HandleApplyPack applies a policy pack to the live bundle.
func (pr *PolicyRoutes) HandleApplyPack(packID string) (int, interface{}) {
	if packID == "" {
		return http.StatusBadRequest, map[string]string{
			"error":   "missing_pack_id",
			"message": "Pack ID is required.",
		}
	}

	pr.Lock()
	defer pr.Unlock()

	added, skipped := policy.ApplyPack(pr.bundle, packID)
	if added == nil && skipped == nil {
		return http.StatusNotFound, map[string]string{
			"error":   "pack_not_found",
			"message": "No policy pack with ID '" + packID + "' found.",
		}
	}
	if pr.saveBundleFn != nil {
		if err := pr.saveBundleFn(*pr.bundle); err != nil {
			return http.StatusInternalServerError, map[string]string{
				"error":   "persist_failed",
				"message": "Pack applied in memory but failed to persist policy bundle: " + err.Error(),
			}
		}
	}

	return http.StatusOK, map[string]interface{}{
		"pack_id":     packID,
		"added":       added,
		"skipped":     skipped,
		"total_rules": len(pr.bundle.Rules),
	}
}

// HandleCreatePack registers a custom policy pack.
func (pr *PolicyRoutes) HandleCreatePack(body []byte) (int, interface{}) {
	var pack policy.PolicyPack
	if err := json.Unmarshal(body, &pack); err != nil {
		return http.StatusBadRequest, map[string]string{
			"error":   "invalid_json",
			"message": "Failed to parse policy pack: " + err.Error(),
		}
	}

	if pack.ID == "" {
		return http.StatusBadRequest, map[string]string{
			"error":   "missing_pack_id",
			"message": "Pack id is required.",
		}
	}

	// Check for duplicates.
	if existing := policy.GetPack(pack.ID); existing != nil {
		return http.StatusConflict, map[string]string{
			"error":   "duplicate_pack_id",
			"message": "A pack with ID '" + pack.ID + "' already exists.",
		}
	}

	policy.AddCustomPack(pack)
	if pr.savePacksFn != nil {
		if err := pr.savePacksFn(policy.GetCustomPacks()); err != nil {
			return http.StatusInternalServerError, map[string]string{
				"error":   "persist_failed",
				"message": "Pack created in memory but failed to persist: " + err.Error(),
			}
		}
	}

	return http.StatusCreated, map[string]interface{}{
		"pack_id": pack.ID,
		"created": true,
	}
}
